`turbo prune --docker` drops pnpm v9 lockfile nodes required by overrides and exact-version peer specs — pruned lockfile fails `pnpm install --frozen-lockfile`

### Verify canary release

- [x] I verified that the issue exists in the latest Turborepo canary release — reproduced on `2.9.10`, `2.9.16`, and `2.9.17-canary.3` (stable/canary pruned outputs are byte-identical, shasum-verified).

### Reproduction

**https://github.com/attehuhtakangas/turbo-prune-pnpm11-lockfile-repro** — `pnpm install && ./repro.sh` demonstrates this issue and #13028 for two prune targets. The full workspace passes `pnpm install --frozen-lockfile`; both pruned outputs fail.

### Environment

- **turbo**: 2.9.10, 2.9.16, 2.9.17-canary.3
- **pnpm**: 11.2.2, `lockfileVersion: '9.0'`

### What happened

pnpm 11 records the **override-resolved value as the importer `specifier:`** (not the `package.json` range) when a `pnpm-workspace.yaml` override applies to a workspace package's dependency. `turbo prune --docker` copies the importer entry verbatim into the pruned lockfile, but **drops the corresponding `packages:`/`snapshots:` entries**, so `pnpm install --frozen-lockfile` against the pruned output fails with `ERR_PNPM_LOCKFILE_MISSING_DEPENDENCY`.

In the repro, with:

```yaml
overrides:
  ms@>=2.1.0: 2.0.0
  picomatch@>=3.0.0 <4: 3.0.2
```

**Case A — override redirects out of the declared range** (`apps/web-override` declares `ms: ^2.1.3`):

```yaml
# pnpm-lock.yaml (full workspace, written by pnpm itself)
importers:
  apps/web-override:
    dependencies:
      ms:
        specifier: 2.0.0     # ← rewritten by pnpm to the override value (declared: ^2.1.3)
        version: 2.0.0
```

The pruned lockfile for `web-override` keeps this importer entry but ends with `snapshots: {}` — the `ms@2.0.0` node is gone:

```
ERR_PNPM_LOCKFILE_MISSING_DEPENDENCY  Broken lockfile: no entry for 'ms@2.0.0' in pnpm-lock.yaml
```

**Case B — override pins within the declared range** (`apps/web-inrange` declares `picomatch: ^3.0.1`):

```yaml
importers:
  apps/web-inrange:
    dependencies:
      picomatch:
        specifier: 3.0.2     # ← rewritten (declared: ^3.0.1; 3.0.2 satisfies it)
        version: 3.0.2
```

```
ERR_PNPM_LOCKFILE_MISSING_DEPENDENCY  Broken lockfile: no entry for 'picomatch@3.0.2' in pnpm-lock.yaml
```

Case B shows this is not about out-of-range redirects: the resolution satisfies the declared range, but the rewritten lockfile specifier (`3.0.2` ≠ `^3.0.1`) appears to defeat prune's importer-dependency matching, and the node's entries are silently omitted.

### Expected behavior

The pruned lockfile passes `pnpm install --frozen-lockfile` like the full lockfile does — override-affected importer dependencies keep their `packages:`/`snapshots:` entries.

### Notes

- **Correction from the original report**: I initially attributed one production failure (`picomatch@3.0.2`) to exact-version `peerDependencies` metadata rewritten by overrides. Bisection disproved that — removing the peer spec left the error intact, and a workspace package's own `picomatch: ^3.0.1` dependency (importer mechanism above) was the actual source. The peer-spec framing is withdrawn; everything reduces to the importer-specifier mechanism demonstrated in the repro.
- In our production monorepo (133 packages, ~5k-entry lockfile) the same mechanism drops 30+ entries per pruned app (e.g. a scoped override `react-markdown@>=9.0.2: 9.0.1` against a declared `^9.0.1`). We work around it with a post-prune script that grafts the missing entries verbatim from the root lockfile.
- Related closed issues in the same area: #11844, #12002, #12013.

--- a/crates/turborepo-lockfiles/src/pnpm/data.rs
+++ b/crates/turborepo-lockfiles/src/pnpm/data.rs
@@ -462,9 +462,9 @@ impl PnpmLockfile {
             return Ok(Some(specifier));
         }
 
-        if resolved_specifier == override_specifier {
-            Ok(Some(resolved_version))
-        } else if self.has_package_by_parts(name, resolved_version, key_buf) {
+        if resolved_specifier == override_specifier
+            || self.has_package_by_parts(name, resolved_version, key_buf)
+        {
             Ok(Some(resolved_version))
         } else if self.has_package_by_parts(name, override_specifier, key_buf) {
             Ok(Some(override_specifier))

--- a/crates/turborepo-lockfiles/src/pnpm/data.rs
+++ b/crates/turborepo-lockfiles/src/pnpm/data.rs
@@ -464,6 +464,8 @@ impl PnpmLockfile {
 
         if resolved_specifier == override_specifier {
             Ok(Some(resolved_version))
+        } else if self.has_package_by_parts(name, resolved_version, key_buf) {
+            Ok(Some(resolved_version))
         } else if self.has_package_by_parts(name, override_specifier, key_buf) {
             Ok(Some(override_specifier))
         } else {
@@ -1136,6 +1138,64 @@ importers:
             !pruned_lockfile.importers.contains_key("apps/docs"),
             "pruned lockfile should still trim workspaces from the final document"
         );
+    }
+
+    #[test]
+    fn test_pnpm_v9_resolves_override_rewritten_importer_versions() {
+        let yaml = r#"lockfileVersion: '9.0'
+
+settings:
+  autoInstallPeers: false
+  excludeLinksFromLockfile: false
+
+overrides:
+  ms@>=2.1.0: 2.0.0
+  picomatch@>=3.0.0 <4: 3.0.2
+
+importers:
+
+  .: {}
+
+  apps/web-inrange:
+    dependencies:
+      picomatch:
+        specifier: 3.0.2
+        version: 3.0.2
+
+  apps/web-override:
+    dependencies:
+      ms:
+        specifier: 2.0.0
+        version: 2.0.0
+
+packages:
+
+  ms@2.0.0:
+    resolution: {integrity: sha512-abc}
+
+  picomatch@3.0.2:
+    resolution: {integrity: sha512-def}
+
+snapshots:
+
+  ms@2.0.0: {}
+
+  picomatch@3.0.2: {}
+"#;
+
+        let lockfile = PnpmLockfile::from_bytes(yaml.as_bytes()).unwrap();
+
+        let in_range = lockfile
+            .resolve_package("apps/web-inrange", "picomatch", "^3.0.1")
+            .unwrap()
+            .expect("override-rewritten in-range specifier should resolve");
+        assert_eq!(in_range.key, "picomatch@3.0.2");
+
+        let out_of_range = lockfile
+            .resolve_package("apps/web-override", "ms", "^2.1.3")
+            .unwrap()
+            .expect("override-rewritten out-of-range specifier should resolve");
+        assert_eq!(out_of_range.key, "ms@2.0.0");
     }
 
     #[test]

--- a/crates/turborepo-lockfiles/src/pnpm/data.rs
+++ b/crates/turborepo-lockfiles/src/pnpm/data.rs
@@ -462,8 +462,12 @@ impl PnpmLockfile {
             return Ok(Some(specifier));
         }
 
-        if resolved_specifier == override_specifier
-            || self.has_package_by_parts(name, resolved_version, key_buf)
+        if resolved_specifier == override_specifier {
+            Ok(Some(resolved_version))
+        } else if self.has_package_by_parts(name, specifier, key_buf) {
+            Ok(Some(specifier))
+        } else if resolved_specifier == resolved_version
+            && self.has_package_by_parts(name, resolved_version, key_buf)
         {
             Ok(Some(resolved_version))
         } else if self.has_package_by_parts(name, override_specifier, key_buf) {
@@ -1196,6 +1200,71 @@ snapshots:
             .unwrap()
             .expect("override-rewritten out-of-range specifier should resolve");
         assert_eq!(out_of_range.key, "ms@2.0.0");
+    }
+
+    #[test]
+    fn test_pnpm_keeps_transitive_exact_version_when_importer_has_same_dep() {
+        let yaml = r#"lockfileVersion: 5.4
+
+importers:
+  packages/a:
+    specifiers:
+      ci-info: ^2.0.0
+      is-ci: ^3.0.1
+    dependencies:
+      ci-info: 2.0.0
+      is-ci: 3.0.1
+
+packages:
+  /ci-info/2.0.0:
+    resolution: {integrity: sha512-abc}
+
+  /ci-info/3.7.1:
+    resolution: {integrity: sha512-def}
+
+  /is-ci/3.0.1:
+    resolution: {integrity: sha512-ghi}
+    dependencies:
+      ci-info: 3.7.1
+"#;
+
+        let lockfile = PnpmLockfile::from_bytes(yaml.as_bytes()).unwrap();
+
+        let transitive = lockfile
+            .resolve_package("packages/a", "ci-info", "3.7.1")
+            .unwrap()
+            .expect("transitive exact version should resolve independently of importer range");
+        assert_eq!(transitive.key, "/ci-info/3.7.1");
+
+        let yaml = r#"lockfileVersion: '9.0'
+
+importers:
+  apps/api:
+    dependencies:
+      express:
+        specifier: 4.21.2
+        version: 4.21.2
+
+packages:
+  express@4.21.2:
+    resolution: {integrity: sha512-abc}
+
+  express@5.1.0:
+    resolution: {integrity: sha512-def}
+
+snapshots:
+  express@4.21.2: {}
+
+  express@5.1.0: {}
+"#;
+
+        let lockfile = PnpmLockfile::from_bytes(yaml.as_bytes()).unwrap();
+
+        let transitive = lockfile
+            .resolve_package("apps/api", "express", "5.1.0")
+            .unwrap()
+            .expect("transitive exact version should resolve independently of importer exact dep");
+        assert_eq!(transitive.key, "express@5.1.0");
     }
 
     #[test]

--- a/lockfile-tests/fixtures/pnpm-11-flat-patch/meta.json
+++ b/lockfile-tests/fixtures/pnpm-11-flat-patch/meta.json
@@ -3,5 +3,6 @@
   "packageManagerVersion": "pnpm@11.0.0-rc.5",
   "lockfileName": "pnpm-lock.yaml",
   "frozenInstallCommand": ["pnpm", "install", "--frozen-lockfile"],
+  "docker": true,
   "expectedFailures": []
 }

--- a/lockfile-tests/fixtures/pnpm-11-flat-patch/packages/b/package.json
+++ b/lockfile-tests/fixtures/pnpm-11-flat-patch/packages/b/package.json
@@ -1,0 +1,5 @@
+{
+  "name": "b",
+  "version": "0.0.0",
+  "private": true
+}

--- a/lockfile-tests/fixtures/pnpm-override-and-patches/apps/web-inrange/package.json
+++ b/lockfile-tests/fixtures/pnpm-override-and-patches/apps/web-inrange/package.json
@@ -1,0 +1,8 @@
+{
+  "name": "web-inrange",
+  "version": "0.0.0",
+  "private": true,
+  "dependencies": {
+    "picomatch": "^3.0.1"
+  }
+}

--- a/lockfile-tests/fixtures/pnpm-override-and-patches/apps/web-override/package.json
+++ b/lockfile-tests/fixtures/pnpm-override-and-patches/apps/web-override/package.json
@@ -1,0 +1,8 @@
+{
+  "name": "web-override",
+  "version": "0.0.0",
+  "private": true,
+  "dependencies": {
+    "ms": "^2.1.3"
+  }
+}

--- a/lockfile-tests/fixtures/pnpm-override-and-patches/meta.json
+++ b/lockfile-tests/fixtures/pnpm-override-and-patches/meta.json
@@ -1,0 +1,7 @@
+{
+  "packageManager": "pnpm",
+  "packageManagerVersion": "pnpm@11.2.2",
+  "lockfileName": "pnpm-lock.yaml",
+  "pruneTargets": ["web-inrange", "web-override"],
+  "docker": true
+}

--- a/lockfile-tests/fixtures/pnpm-override-and-patches/package.json
+++ b/lockfile-tests/fixtures/pnpm-override-and-patches/package.json
@@ -1,0 +1,5 @@
+{
+  "name": "turbo-prune-pnpm11-lockfile-repro",
+  "private": true,
+  "packageManager": "pnpm@11.2.2"
+}

--- a/lockfile-tests/fixtures/pnpm-override-and-patches/packages/other/package.json
+++ b/lockfile-tests/fixtures/pnpm-override-and-patches/packages/other/package.json
@@ -1,0 +1,8 @@
+{
+  "name": "other",
+  "version": "0.0.0",
+  "private": true,
+  "dependencies": {
+    "is-odd": "3.0.1"
+  }
+}

--- a/lockfile-tests/fixtures/pnpm-override-and-patches/patches/is-odd@3.0.1.patch
+++ b/lockfile-tests/fixtures/pnpm-override-and-patches/patches/is-odd@3.0.1.patch
@@ -1,0 +1,10 @@
+diff --git a/index.js b/index.js
+index 79d1f22a8e7a27efb8841bb83cb682ea1ff3a59c..ce3722059108565de2de0676ff0ffcab2bc5813d 100644
+--- a/index.js
++++ b/index.js
+@@ -23,3 +23,5 @@ module.exports = function isOdd(value) {
+   return (n % 2) === 1;
+ };
+ 
++
++// repro: patched marker

--- a/lockfile-tests/fixtures/pnpm-override-and-patches/pnpm-workspace.yaml
+++ b/lockfile-tests/fixtures/pnpm-override-and-patches/pnpm-workspace.yaml
@@ -1,0 +1,14 @@
+packages:
+  - apps/*
+  - packages/*
+
+autoInstallPeers: false
+dedupePeers: true
+allowUnusedPatches: true
+
+overrides:
+  ms@>=2.1.0: 2.0.0
+  picomatch@>=3.0.0 <4: 3.0.2
+
+patchedDependencies:
+  is-odd@3.0.1: patches/is-odd@3.0.1.patch

--- a/lockfile-tests/fixtures/pnpm-override-and-patches/turbo.json
+++ b/lockfile-tests/fixtures/pnpm-override-and-patches/turbo.json
@@ -1,0 +1,6 @@
+{
+  "$schema": "https://turborepo.com/schema.json",
+  "tasks": {
+    "build": {}
+  }
+}

